### ISSUE TITLE
grcov reports zero coverage for a test that *did* run

### ISSUE THREAD

**Issue by Minoru** [mozilla/grcov#438]
This is a weird regression that I only see on Travis. I couldn't reproduce this locally.

After I replaced `-Zno-landing-pads` as per #428, Coveralls started showing me zero coverage for a test in an otherwise-covered module: https://coveralls.io/builds/30658558/source?filename=rust%2Fstrprintf%2Fsrc%2Fspecifiers_iterator.rs#L116 Yet Travis tells me that the test *did* run, and finished ok: https://travis-ci.com/github/newsboat/newsboat/jobs/330683115#L1438

An earlier Coveralls report with Nightly 2020-04-29 and `-Zno-landing-pads` showed this module (and the contained test) as fully covered: https://coveralls.io/builds/30634536/source?filename=rust/strprintf/src/specifiers_iterator.rs

I tried to reproduce this locally with `-t lcov` and `genhtml`, but report shows the module as fully covered.

Any ideas on what might be causing this, and how I might pinpoint the issue? I can understand if some line becomes not-covered because newer compiler optimizes it away, but for a whole test to disappear is pretty weird.

**Comment by marco-c** [comment]
> I tried to reproduce this locally with -t lcov and genhtml, but report shows the module as fully covered.

With the same options (-Cpanic=abort and so on) you are using on CI?

**Comment by calixteman** [comment]
I can't reproduce the error locally.
And https://coveralls.io/builds/30658558/source?filename=rust%2Fstrprintf%2Fsrc%2Fspecifiers_iterator.rs#L48 gives the impress that the test really didn't run.
If you manage to reproduce consistently the issue on CI, maybe you could add a `println!(...)` in the test and add `-- --nocapture` to `cargo test` to check if the test is really ran.
It's very weird.

**Comment by Minoru** [comment]
> With the same options (-Cpanic=abort and so on) you are using on CI?

Yes.

> If you manage to reproduce consistently the issue on CI, maybe you could add a println!(...) in the test and add -- --nocapture to cargo test to check if the test is really ran.

[Done](https://github.com/newsboat/newsboat/pull/924/commits/b1cee51c4bbe8160ac0910a41c0be01a1feb8e96#diff-1fdaecbe7d078fcd7aa6144ca0e19d1eR117). The test [definitely ran](https://travis-ci.com/github/newsboat/newsboat/jobs/331917717#L1503-L1505), and [the coverage went back up](https://coveralls.io/jobs/62513827/source_files/449610260#L116) =\

I then reverted my test commit and ran the job again, and [despite the test showing up in the output](https://travis-ci.com/github/newsboat/newsboat/jobs/331925584#L1433), the coverage [went down again](https://coveralls.io/builds/30701963/source?filename=rust/strprintf/src/specifiers_iterator.rs).

I then added `-- --test-threads=1`, and [the coverage went back up](https://coveralls.io/builds/30702892/source?filename=rust/strprintf/src/specifiers_iterator.rs). So the problem might not be in grcov, and is instead in Rust's test runner? Should I submit a bug to the rustc repo?

**Comment by calixteman** [comment]
I think you should file a bug on cargo instead I'm pretty sure that the test didn't run which is pretty bad (if I'm correct of course).

**Comment by Minoru** [comment]
While preparing a report for cargo repo, I noticed that `-Cinline-threshold` was replaced by `-Copt-level` in the README here, so I applied these changes to my repo as well, and [the issue went away](https://coveralls.io/builds/30705976/source?filename=rust/strprintf/src/specifiers_iterator.rs). Since #411 doesn't contain any rationale, I have no idea if it's just a coincidence, or if `opt-level` implies some other options that are related to this.

Anyway, the problem is fixed now, so I won't file anything with Cargo, and I'm closing this now. Thanks for support, and sorry for bothering you with something that isn't even grcov's issue! :heart:
